### DISCUSSION TITLE
Detekt Gradle plugin

### DISCUSSION THREAD

**Discussion by picklebento** [detekt/detekt#4052] [General]
There are recent discussions with respect to the detekt-gradle-plugin. I would like to formally reach a consensus that if we want to build detekt-gradle-plugin as...
- Subproject
   - As mentioned in https://github.com/detekt/detekt/pull/2493, verifying breaking changes in detekt-cli is easier. For instance, adding a new cli parameters
- Composite-build
   - Verifying gradle plugin changes is easier. We only need to run ./gradlew detekt in our root project.
   - Build speed is increased since detekt-gradle-plugin need to be evaluated every time in the initialization / configuration phase.
- Separate project such as detekt-compiler-plugin
   - Both scenarios became more difficult.

I assume we change gradle plugin more often than adding cli parameters, therefore, moving it to composite-build is making us more productive. Any other thoughts?

**Comment by BraisGabin** [comment]
Would any of those help with #2957? I don't have any strong opinion here. If we had a really stable API I would vote for the third option. But we don't so 1 or 2 are fine for me.

**Comment by cortinico** [comment]
I'm in for the composite build approach (I'm biased maybe :)). I believe it would help us ship a better product overall.

The problem is that currently detekt `main` runs on the "latest" released version of the Gradle plugin. This creates discrepancy, especially during the release phase. We have PRs that are touching the Gradle plugin code, maybe they not well tested...and we end up discovering problems during the release phase.

To mitigate the build impact, we could opt for an hybrid approach. Like we could have an env variable that optionally imports the composite build (for developers that wants to develop for it and for a CI workflow), and as an alternative fallsback to getting the artifact of the Gradle plugin from Gradle Plugin Central.
